symex-evil.el

```diff
@@ -47,7 +47,8 @@
   :tag " <λ> "
   :message "-- SYMEX --"
   :enable (normal)
-  :exit-hook (symex-exit-mode))
+  :entry-hook (evil-repeat-stop)
+  :exit-hook (evil-repeat-start-hook symex-exit-mode))
```

Comment:
What was the problem you were seeing with insertion? Earlier I tried adding commands like `symex-insert-before` to `symex--evil-repeatable-commands` and they seemed to work, though I didn't test every command. I believe evil treats (entry into insert state + entering text + exit) as a single command, so in principle that should work. Starting to record upon leaving Symex state sounds like it could be risky.

Reply:
I might have misunderstood the expected behavior here, but this is trying to follow the same pattern used for normal state, I.e. all edits done outside of symex are treated as a single repeat unit.

Evil does this by started to record during the exit hook when leaving normal, but no special action needs to be taken when returning to normal state because that's the only state which does any recording by default. I'm turning off recording on entry into symex to account for the case where a user moves from normal into symex.

If the expected behavior is to just repeat the command which took the buffer into insert state, while not keeping track of edits, then I think everything was working before (except that the functions weren't marked repeatable).

Reply:
Interesting, yeah that's the expected behavior. On my end, it actually works correctly in the previous version that did not have these changes, after adding the insertion commands to `symex--evil-repeatable-commands`. Check it out, here are two GIFs, the first one _before_ the latest changes, and the second one including the latest changes.

20d0249513b624622e94eb56e8be673b854fdb00 + adding the leftover insertion commands:
<details>
<img src="https://user-images.githubusercontent.com/401668/175107248-4905e8db-1430-4281-ab25-2858adf21228.gif">
</details>

c468610c1c95d67eea46bc53d5cc033dac18e76c:

<details>
<img src="https://user-images.githubusercontent.com/401668/175107278-7104a380-0b75-4078-91aa-31d4275adbdb.gif">
</details>

In the second case, it changes to Normal state bindings after returning to Symex state, and I need to cycle to Normal->Symex again to restore the symex bindings. Is the behavior different on your end? If so, this might warrant a closer look at our config to see what's different.

Reply:
Looks like I am seeing the same behavior on my end--but had assumed this was because I had broken my rigpa setup. I did some reverting to put this branch into the state of your first GIF; although, I don't think I fully understand how the insertions are being recorded into the repeat ring. I'll keep poking around and see if I had internalize what's happening here  ¯\\_(ツ)_/¯

Reply:
It's good that we're seeing the same behavior at least, that takes away some uncertainty 🙂 

And good call, it is odd that it still works even though recording is not enabled like for Normal state, and it would be nice to understand that. [This comment](https://github.com/emacs-evil/evil/blob/cc55fa633d4f726953654326287abb9a233e0c90/evil-repeat.el#L52-L54) might have something to do with it. I added a print statement in [`evil-repeat-start-hook`](https://github.com/emacs-evil/evil/blob/cc55fa633d4f726953654326287abb9a233e0c90/evil-repeat.el#L258) and another in [`evil-repeat-pre-hook`](https://github.com/emacs-evil/evil/blob/cc55fa633d4f726953654326287abb9a233e0c90/evil-repeat.el#L267). Looks like the former is triggered on `i` in Normal state, but not for `i` in Symex state, but the latter is triggered in Symex state but not in Normal state. That may have something to do with why it still ends up recording the repeatable actions in insert mode even though it isn't configured to start recording on exit the same way that Normal state is. But I still don't quite understand how it works. I'll take another look soon.

Reply:
Ok, I looked at `evil-repeat-start` using [edebug](https://github.com/countvajhula/symex.el/blob/master/DSL-Docs.rst#using-a-debugger-edebug), and this is the backtrace when you type `i` in Symex state:

```
  evil-repeat-start()
  symex-evil-repeat-start-recording-advice()
  apply(symex-evil-repeat-start-recording-advice nil)
  evil-repeat-pre-hook()
```

I'm not entirely sure how `evil-repeat-pre-hook` knows to be triggered by commands like `symex-insert-at-beginning`, but other than that, this looks about right to me -- the advice function you added is being triggered by `evil-repeat-pre-hook` as described in [this comment](https://github.com/emacs-evil/evil/blob/cc55fa633d4f726953654326287abb9a233e0c90/evil-repeat.el#L52-L54), and then `symex-evil-repeat-start-recording-advice` calls `evil-repeat-start`.

Wdyt?

If you're convinced, then I think this PR is just about ready to merge 😄

Reply:
Oh snap, that does make sense to me now. I haven't noticed any problems using this branch these past couple of weeks, so if you feel ready to merge I say go for it!